rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,608 @@
+    0011 -> SE
+
+We need a set of rules for:
+
+ 1. Where merged entries are to be inserted into the table
+ 2. Which merges are allowed
+
+"Ordered Covering"
+==================
+
+The algorithm implemented here, "Ordered Covering", provides the following
+rules:
+
+ 1. Routing tables are to be kept sorted in increasing order of "generality",
+    that is the number of ``X``s in the entry. An entry with the key-mask pair
+    ``00XX`` must be placed below any entries with fewer ``X``s in their
+    key-mask pairs (e.g., below ``0000`` and ``000X``).
+
+    a) New entries must also be inserted below any entries of the same
+       generality. If ``XX00`` were already present in the table the new entry
+       ``0XX0`` must be inserted below it.
+
+ 2. The only merges allowed are those which:
+
+    a) would not cause one of the entries in the merge to be "hidden" below
+       an entry of lesser generality than the merged entry but which matched
+       any of the same keys. For example, merging ``0010`` and ``0001`` would
+       not be allowed if the new entry would be placed below the existing
+       entry ``000X`` as this would "hide" ``0001``.
+    b) would not cause an entry "contained" within an entry of higher
+       generality to be hidden by the insertion of a new entry. For example, if
+       the entry ``XXXX`` had been formed by merging the entries ``0011`` and
+       ``1100`` then merging of the entries ``1101`` and ``1110`` would not be
+       allowed as it would cause the entry ``11XX`` to be inserted above
+       ``XXXX`` in the table and would hide ``1100``.
+
+Following these rules ensures that the minimised table will be functionally
+equivalent to the original table provided that the original table was invariant
+under reordering OR was provided in increasing order of generality.
+"""
+from collections import defaultdict, namedtuple
+from rig.routing_table import RoutingTableEntry
+from six import itervalues
+
+
+def minimise(routing_table, target_length=1023):
+    """Reduce the size of a routing table by merging together entries where
+    possible.
+
+    .. warning::
+
+        The minimised table may be larger than `target_length`, it is
+        important to check the size of the returned table.
+
+    .. warning::
+
+        It is assumed that the input routing table is not in any particular
+        order and may be reordered into ascending order of generality without
+        affecting routing correctness.  It is also assumed that if this table
+        is unordered it is at least orthogonal (i.e., there are no two entries
+        which would match the same key) -- this assumption is not checked!
+
+        .. note::
+
+            Use :py:meth:`~rig.routing_table.expand_entries` to generate an
+            orthogonal table and receive warnings if the input table were not
+            orthogonal.
+
+    .. warning::
+
+        The input routing table *must* also include entries which could be
```

Comment:
This should really be the first warning. I suspect most people will be producing orthogonal tables without thinking (if I understand rightly, the routing table generation function already does this, right?) but omitting default routes is a really easy way to very confusingly shoot yourself in the foot since the results can clearly work *some* of the time...